File: HelpBot/src/main/java/com/example/HelpBot/web/HelpBotJsCommand.java
package com.example.HelpBot.web;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;

/**
 * HelpBot Web SDK 命令封装（生成 evaluateJavascript 可执行的 JS 字符串）
 * 安全性：
 * - 所有字符串入参使用 JSONObject.quote(...) 做 JS 字符串转义，避免注入与语法错误。
 */
public final class HelpBotJsCommand {

    private HelpBotJsCommand() {
        super();
    }

    @NonNull
    public static String buildInitConfig(@NonNull final JSONObject config) {
        return "window.HelpBotConfig = " + config.toString() + ";";
    }

    @NonNull
    public static String buildLoadScript(@NonNull final String jsUrl) {
        final String safeUrl = JSONObject.quote(jsUrl);
        return "var newScript=document.createElement(\"script\");newScript.src=" + safeUrl + ";document.body.appendChild(newScript);";
    }

    @NonNull
    public static String buildSetTokenAndConnect(@NonNull final String token) {
        return "HelpBot('setTokenAndConnect', " + JSONObject.quote(token) + ");";
    }

    @NonNull
    public static String buildOpen() {
        return "HelpBot('open');";
    }

    @NonNull
    public static String buildClose() {
        return "HelpBot('close');";
    }

    @NonNull
    public static String buildToggle() {
        return "HelpBot('toggle');";
    }

    @NonNull
    public static String buildDestroy() {
        return "HelpBot('destroy');";
    }

    @NonNull
    public static String buildUpdateUserSdkMeta(@NonNull final JSONObject meta) {
        return "HelpBot('updateUserSdkMeta', " + meta.toString() + ");";
    }

    @NonNull
    public static String buildUpdateUserMeta(@NonNull final JSONObject meta) {
        return "HelpBot('updateUserMeta', " + meta.toString() + ");";
    }

    @NonNull
    public static String buildAddIssueTags(@NonNull final JSONArray tags) {
        return "HelpBot('addIssueTags', " + tags.toString() + ");";
    }

    @NonNull
    public static String buildRemoveIssueTags(@NonNull final JSONArray tags) {
        return "HelpBot('removeIssueTags', " + tags.toString() + ");";
    }

    @NonNull
    public static String buildGetStatus(@Nullable final String callbackFnName) {
        // 如果宿主提供回调名，把 status 传回 window[callback](status)
        if (callbackFnName == null || callbackFnName.trim().isEmpty()) {
            return "HelpBot('getStatus');";
        }
        final String safeCb = callbackFnName.trim();
        return "try{var s=HelpBot('getStatus');if(window['" + safeCb + "']){window['" + safeCb + "'](s);}}catch(e){}";
    }
}

File: SDK/iOS/HelpBotSDK/Sources/HelpBotSDK/Web/HelpBotJsCommand.swift
import Foundation

/**
 HelpBot Web SDK 命令封装（生成 evaluateJavaScript 可执行的 JS 字符串）。

 安全性：
 - 所有字符串入参使用 JSON 编码生成安全字符串字面量，避免注入与语法错误。
 */
enum HelpBotJsCommand {
    static func buildInitConfig(_ config: [String: Any]) -> String {
        let json = HelpBotJsonUtils.toJsonString(config) ?? "{}"
        return "window.HelpBotConfig = \(json);"
    }

    static func buildLoadScript(_ jsUrl: String) -> String {
        let safeUrl = quoteJsString(jsUrl)
        return "var newScript=document.createElement(\"script\");newScript.src=\(safeUrl);document.body.appendChild(newScript);"
    }

    static func buildSetTokenAndConnect(_ token: String) -> String {
        return "HelpBot('setTokenAndConnect', \(quoteJsString(token)));"
    }

    static func buildOpen() -> String { "HelpBot('open');" }
    static func buildClose() -> String { "HelpBot('close');" }
    static func buildToggle() -> String { "HelpBot('toggle');" }
    static func buildDestroy() -> String { "HelpBot('destroy');" }

    static func buildUpdateUserSdkMeta(_ meta: [String: Any]) -> String {
        let json = HelpBotJsonUtils.toJsonString(meta) ?? "{}"
        return "HelpBot('updateUserSdkMeta', \(json));"
    }

    static func buildUpdateUserMeta(_ meta: [String: Any]) -> String {
        let json = HelpBotJsonUtils.toJsonString(meta) ?? "{}"
        return "HelpBot('updateUserMeta', \(json));"
    }

    static func buildAddIssueTags(_ tags: [String]) -> String {
        let json = HelpBotJsonUtils.toJsonString(tags) ?? "[]"
        return "HelpBot('addIssueTags', \(json));"
    }

    static func buildRemoveIssueTags(_ tags: [String]) -> String {
        let json = HelpBotJsonUtils.toJsonString(tags) ?? "[]"
        return "HelpBot('removeIssueTags', \(json));"
    }

    static func buildShowFAQs(_ config: [String: Any]) -> String {
        let json = HelpBotJsonUtils.toJsonString(config) ?? "{}"
        return "HelpBot('showFAQs', \(json));"
    }

    static func buildGetStatus(callbackFnName: String?) -> String {
        guard let callbackFnName, !callbackFnName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return "HelpBot('getStatus');"
        }
        let cb = callbackFnName.trimmingCharacters(in: .whitespacesAndNewlines)
        return "try{var s=HelpBot('getStatus');if(window['\(cb)']){window['\(cb)'](s);}}catch(e){}"
    }

    /// 生成 JS 字符串字面量（带引号）
    private static func quoteJsString(_ raw: String) -> String {
        // 利用 JSON 序列化保证转义正确：["xxx"] -> "xxx"
        guard let json = HelpBotJsonUtils.toJsonString([raw]) else { return "\"\"" }
        if json.count >= 4, json.hasPrefix("[\""), json.hasSuffix("\"]") {
            return String(json.dropFirst(1).dropLast(1))
        }
        return "\"\""
    }
}


